### DISCUSSION TITLE
When saving/naming a file, how do you advance caracter without a rotary encoder ?

### DISCUSSION THREAD

**Discussion by Onosendaifr** [probonopd/MiniDexed#691] [Q&A]
Maybe i miss something but i can't find how to advance the cursor of my caracter after i'm done with the first one. I've tried all the MIDI button i've mapped, does that mean it's not possible without rotary encoder yet ?

**Comment by Onosendaifr** [comment]
Did one of you succeed to name a file withtout a rotary encoder ? Only with MIDI buttons ?

**Comment by diyelectromusic** [comment]
Is this trying to save a performance? If so, then this doesn't look hopeful... https://github.com/probonopd/MiniDexed/blob/main/src/uimenu.cpp#L1655

Kevin

**Reply by diyelectromusic** [reply]
> Is this trying to save a performance? If so, then this doesn't look hopeful... https://github.com/probonopd/MiniDexed/blob/main/src/uimenu.cpp#L1655
> 
> Kevin

Ok, ignore that - actually it uses the "shortcut" action of the encoder which isn't implemented for buttons, so I'm not quite sure how that could work with the code being as it is I'm afraid...

I guess I might be able to implement some kind of "hold select whilst using Prev/Next" type action, but I'm not quite sure how that could work atm.

Kevin

**Reply by Onosendaifr** [reply]
I'm ready to try :)

**Reply by diyelectromusic** [reply]
Feel free, but the interactions of buttons/encoder/menus is a bit... er... odd (in my view).

Kevin

**Reply by Onosendaifr** [reply]
Better having a solution than nothing, because adding an analog rotary controller is not as easy as pluging a USB MIDI controller. 

By the way, why is it not possible to add a MIDI rotary knob for this ?

**Reply by diyelectromusic** [reply]
I'm not saying anything is not possible, I'm just relaying what is in the code right now.

Anything else requires effort to figure out, write, and test a code update :)

Aside: how does a MIDI rotary knob work?  MIDI tends to do CCs with a specific value (0..127) so doesn't really include an encoder-like operation.  It is of course possible to have a "going up/down" type operation, but that doesn't really take advantage of a knob of course - that is really just two buttons.

Kevin

**Reply by Onosendaifr** [reply]
Can't say i'm not a technician nor coder :) But they are pretty commun on a MIDI controller.

For exemple i use this : https://store.djtechtools.com/products/midi-fighter-twister

You can also push them and everything send CC.

**Reply by diyelectromusic** [reply]
Ah I see - that makes sense.  The encoder is two-way, the parameters are sent to the encoder for indicating the "position" and then adjustments are sent back.  A bit like those motorised volume controls you used to be able to get on higher end stereos back in the day (that looked really cool!).

These need proper DAW support though to do their thing and that isn't so easy in a multi-MIDI channel setup like MiniDexed as we have to automatically send out value update as well as receive them to keep everything in sync and that can mess with other things.

It's not impossible... but it isn't easy :)

Of course the original question is about driving the UI not about adjusting CC values, but then part of the issue is that our parameters aren't all "CC-able".  So maybe that is the way to think about it... finally just go for it and "CC all the things"

Kevin
